### PR TITLE
docs/testplan: Adding a testplan for Server Side Copy

### PR DESCRIPTION
**Pull request by nadimhq**
This is a testplan for Server Side Copy that follows the testplan template.

 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

**Review by littleskunk — DISMISSED** [review]

File: docs/testplan/server-side-copy-testplan.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+|                        | Test Impact on Metainfo > Expired Deletion                         | While metainfo loop is going through expired deletion, it should be possible to server side copy with no negative impact on former services                                                                                                                             |          |
+|                        | Test Impact on Accounting/Tally                                    | While accounting services are running, it should be possible to server side copy with no negative impact on former services                                                                                                                                             |          |
+|                        | Test Expiration Dates                                              | For server side copy users are allowed to make new copies of pieces which allows users to change their expiration dates, so in this way users are able to change expiration date of the new object through server side copy                                             |          |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:21]: This won't be possible as currently designed, as storage nodes will delete the pieces on the original expiration date.

Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: Updated comment section of testplan with your reasoning, Fadila!

File: docs/testplan/server-side-copy-testplan.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+|                        | Copy inline segment                                           | User should copy inline segment without any errors                                                                                                                                                                                                                      |          |
+|                        | Copy remote segment                                           | User should copy inline segment without any errors                                                                                                                                                                                                                      |          |
+|                        | Start server-side copying process and drop connection         | File should be copied even when connection is dropped                                                                                                                                                                                                                   |          |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:53]: I'm confused "drop connection". There will be first a call to BeginCopyObject satellite endpoint, and then a call to FinishCopyObject. If the connection is dropped before FinishCopyObject is called, the object won't be copied.

Maybe the confusion comes from me using the term "atomic operation". It was more from the satellite perspective, in the sense that we should not partially copied objects in the db.

File: docs/testplan/server-side-copy-testplan.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+|                        | Copy 10000 Or More Parts                                   | If a user uploads an object with 10000 parts or more and server side copies it from the source object to the destination object, it should be possible                                                                                                               |          |
+|                        | Copy 10000 Or More Segments                                   | If a user uploads an object with 10000 segments or more and server side copies it from the source object to the destination object, it should be possible                                                                                |          |
+|                        | Copy Objects More than 5GB                                    | If objects are more than 5GB in size, then upload-part-copy api must be used for server side copy                                                                                                                                                       |          |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:48]: From the satellite perspective, I don't think we will have a limit on size of the object, the limit will only on segment count.

File: docs/testplan/server-side-copy-testplan.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+|                        | ancestor_stream_id negative                                   | If a segment hasn't been copied, then it is empty under ancestor_stream_id in the db                                                                                                                                                                                    |          |
+|                        | ancestor_stream_id positive                                   | If a segment has been copied, then for the copied segment it contains the ancestor_stream_id which is the stream_id for the original segment                                                                                                                            |          |
+|                        | reference_counter integer                                     | there is a reference_counter integer which counts the number of segments having stream_id as ancestor_stream_id, so If a new segment has been server side copied then the integer should reflect that (+1)                                                              |          |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:25]: We should add the edge case where we already have max(reference_counter) copies. We should throw an error in such case.

File: docs/testplan/server-side-copy-testplan.md (hunk 3 above)
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: Updated

File: docs/testplan/server-side-copy-testplan.md (hunk 4 above)
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: Updated

File: docs/testplan/server-side-copy-testplan.md (hunk 2 above)
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: Updated

File: docs/testplan/server-side-copy-testplan.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+|                        | Object Deleted During Copy- remote segment                    | If a user uploads an remote segment and then server side copies from the source object to the destination object all while deleting the original source object, then an error is thrown                                                                                 |          |
+|                        | Object Deleted After Copy- remote segment                     | If a user uploads an remote segment, server side copies it from the source object to the destination object and then deletes the original source object, then an error is not shown                                                                                     |          |
+|                        | Copy 10000 Or More Parts                                   | If a user uploads an object with 10000 parts or more and server side copies it from the source object to the destination object, it should be possible                                                                                                               |          |
```
[the remark this answers is not in the document]
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: Updated

File: docs/testplan/server-side-copy-testplan.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+|                        | Copy 0 bytes object                                           | User should copy 0 bytes object without any error                                                                                                                                                                                                                       |          |
+|                        | Copy inline segment                                           | User should copy inline segment without any errors                                                                                                                                                                                                                      |          |
+|                        | Copy remote segment                                           | User should copy inline segment without any errors                                                                                                                                                                                                                      |          |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:53]: "remote" :)

File: docs/testplan/server-side-copy-testplan.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+|                        | ~~Start server-side copying process and drop connection         | File should be copied even when connection is dropped~~                                                                                                                                                                                                                  | If the connection is dropped before FinishCopyObject is called, the object won't be copied         |
```
Comment (Fadila82) [inline docs/testplan/server-side-copy-testplan.md:54]: nit: there are extras "~"

File: docs/testplan/server-side-copy-testplan.md (hunk 6 above)
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: fixed!

File: docs/testplan/server-side-copy-testplan.md (hunk 7 above)
Comment (nadimhq) [reply docs/testplan/server-side-copy-testplan.md]: fixed strikethrough!

**Review by storj-antonio — APPROVED** [review]
Looks good to me, I rebased the repo as it was out of date and added my approval.

**Review by egonelbre — COMMENTED** [review]
Wouldn't this be better as a issue where you can have checkboxes on each issue and update them directly instead of having to go through committing and reviewing?

**Review by storj-antonio — APPROVED** [review]

**Review by littleskunk — APPROVED** [review]

**Pull request merged**